ambry-tools/src/main/java/com.github.ambry/store/CompactionVerifier.java

```diff
@@ -0,0 +1,620 @@
+    UUID incarnationId = UUID.randomUUID();
+    MessageStoreRecovery recovery = new MessageStoreRecovery() {
+      @Override
+      public List<MessageInfo> recover(Read read, long startOffset, long endOffset, StoreKeyFactory factory)
+          throws IOException {
+        return Collections.EMPTY_LIST;
+      }
+    };
+    MessageStoreHardDelete hardDelete = new BlobStoreHardDelete();
+
+    // load "src compaction" log and index
+    srcLog = new Log(srcDir.getAbsolutePath(), verifierConfig.storeCapacity, -1, srcMetrics);
+    srcIndex =
+        new PersistentIndex(srcDir.getAbsolutePath(), null, srcLog, storeConfig, storeKeyFactory, recovery, hardDelete,
+            srcMetrics, SystemTime.getInstance(), sessionId, incarnationId);
+
+    // load "tgt" compaction log and index
+    tgtLog = new Log(tgtDir.getAbsolutePath(), verifierConfig.storeCapacity, -1, tgtMetrics);
+    tgtIndex =
+        new PersistentIndex(tgtDir.getAbsolutePath(), null, tgtLog, storeConfig, storeKeyFactory, recovery, hardDelete,
+            tgtMetrics, SystemTime.getInstance(), sessionId, incarnationId);
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
+      srcIndex.close();
+      tgtIndex.close();
+    } catch (StoreException e) {
+      throw new IOException(e);
+    }
+    srcLog.close();
+    tgtLog.close();
+    cLog.close();
+  }
+
+  /**
+   * Verifies the compaction by checking structure and data integrity.
+   * @throws IOException if there is any I/O error.
+   * @throws StoreException if there is any problem performing store operations.
+   */
+  private void verifyCompaction() throws IOException, StoreException {
+    verifyStructure();
+    verifyData();
+  }
+
+  /**
+   * Verifies that the structure of the store is consistent (both in mem and on disk). See the documentation of the
+   * class for details on the verifications.
+   */
+  private void verifyStructure() {
+    LOGGER.info("Verifying structure");
+    List<String> srcSegmentNames = getSegmentNames(srcLog);
+    List<String> tgtSegmentNames = getSegmentNames(tgtLog);
+    List<String> expectedTgtSegmentNames = new ArrayList<>();
+
+    // 1. Check that log segments that should exist, still exist and ones that should not, no longer do
+    for (String srcSegmentName : srcSegmentNames) {
+      if (segmentsCompactedNames.contains(srcSegmentName)) {
+        assert !tgtSegmentNames.contains(srcSegmentName) :
+            "Compacted log should not contain segment: " + srcSegmentName;
+      } else {
+        assert tgtSegmentNames.contains(srcSegmentName) : "Compacted log is missing segment: " + srcSegmentName;
+        expectedTgtSegmentNames.add(srcSegmentName);
+      }
+    }
+
+    // 2. Check that log segments that are newly created have the right generation number
+    for (String segmentCompactedName : segmentsCompactedNames) {
+      String tgtSegmentName = LogSegmentNameHelper.getNextGenerationName(segmentCompactedName);
```

Comment:
There is no such expectation. Can clarify this offline

Also tested this out when a store is recompacted (twice)